### DISCUSSION TITLE
Support for 'nulls not distinct' on postgres unique indexes

### DISCUSSION THREAD

**Discussion by ladal1** [mikro-orm/mikro-orm#4971] [Ideas]
Ability to use nulls not distinct option on unique indexes for postgres, especially when used with upsert functions, which will currently throw when trying to construct query which would have null in any of the fields from the unique index. 

Proposal basically boils down to this functionality
```ts
@Unique({
  name: 'nulls_not_unique_idx',
  properties: ['account', 'nullable_column'],
  options: {
    nullsNotDistinct: true
  }
})
```

```ts
await em.upsert(
  Entity,
  {
    account: {
      id: args.accountId ?? user?.id,
    },
    nullableColumn: null,
    otherColumn: 'Upserts on nullable_column and account_id, even though the value was null'
  })),
);
```

**Comment by B4nan** [comment]
You can use index expressions to build any kind of index, in v6 they are also supported in the `@Unique()` decorator.

---

I am not entirely happy with the way indexes work now, but some time ago I surrendered and introduced the index expressions instead, as it felt impossible to catch every possible (and dialect/driver-specific) configuration. 

Maybe we could improve the way how the index expressions work, to have something in the middle? Let the ORM to create the base part for the `create index` query and only allow to modify it? Or append some SQL to the end of it?

**Reply by ladal1** [reply]
Building is fine, however upsert will not work as far as my experimentation got, or am I wrong?
We're currently on v6, the latest versions and it's great, the upsert is the main feature here.

**Reply by B4nan** [reply]
Oh, I thought you only wanted a way to generate the index/constraint. So for the upsert we'd have to add some new option to disable the checks? Could you prepare a repro for that so I can understand better what needs to be done?

**Reply by ladal1** [reply]
Sure, I'll quickly make something and test it, I would probably prefer it on the unique decorator itself, as it does affect (at least in postgres) behaviour on all upserts, but simple option on the `.upsert` can work as well

**Reply by ladal1** [reply]
Here is an example of the behaviour
```ts
@Entity()
@Unique({
  name: 'entity_null_not_unique_index,
  properties: ['orderId', 'distinctNullableValue'],
})
class TestEntity {
  @PrimaryKey()
  id!: number
  
  @Property()
  orderId!: string
  
  @Property({ nullable: true})
  distinctNullableValue!: string | null
  
  @Property()
  someNonUniqueValue!: string
}

em.upsert(TestEntity, {
  orderId: '123',
  distinctNullableValue: null,
  someNonUniqueValue: 'Mikro ORM is great',
})
```
This will fail (at least in my testing on only slightly different entity) with
```
Unique property value required for upsert, provide one of: id, orderId + distinctNullableValue
```

**Reply by ladal1** [reply]
I realize I havent specified the index here to be nulls not distinct as works in postgres 15+, I'll replicate it in the unique index maker if I can 😄 

---

Don't have the enviroment setup to test the expressions, but here's the direct postgres query:
```
alter table "test_entity" add constraint "entity_null_not_unique_index" unique nulls not distinct ("orderId", "distinctNullableValue");
```

**Reply by B4nan** [reply]
That's a good point that it affects all the upserts.

Btw since you are on v6, are you using the latest version with the reworked joined strategy? Curious if it didn't break anything (it should rather fix things for those who used it, but you never know).

**Reply by ladal1** [reply]
We didn't have time to test it right now, so we just set the option to the old strategy, but my aim is to test it this week (probably around wednesday, I'll make sure to report anything I find, however we already work with `populateWhere: PopulateHint.INFER` global setting so I'm not sure if it will do much (I've only skimmed the rc2 release notes on slack)

**Comment by hinsxd** [comment]
Hi fellow AI coders, This discussion post is contributing heavily to the hallucination when an agent wants this uniqueness on MikroORM. Please do not fall for the trap.

Conclusion: `nullsNotDistinct` is not a MikroORM feature. Period.

Thank you.
